apps/mymedicare_cb/models.py

```diff
@@ -1,76 +1,106 @@
-
-    # Determine patient_id
-    fhir_source = get_resourcerouter()
-    crosswalk, _ = Crosswalk.objects.get_or_create(
-        user=user, fhir_source=fhir_source)
-    hicn = user_info.get('hicn', "")
-    crosswalk.set_hicn(hicn)
-    crosswalk.save()
+        pass
+
+    first_name = user_info.get('given_name', "")
+    last_name = user_info.get('family_name', "")
+    email = user_info.get('email', "")
 
     try:
-        backend_data = authenticate_crosswalk(crosswalk)
+        fhir_id, backend_data = match_hicn_hash(hicn_hash)
         # Get first and last name from FHIR if not in OIDC Userinfo response.
-        if user_info['given_name'] == "" or user_info['family_name'] == "":
-            if 'entry' in backend_data:
-                if 'name' in backend_data['entry'][0]['resource']:
-                    names = backend_data['entry'][0]['resource']['name']
-                    first_name = ""
-                    last_name = ""
-                    for n in names:
-                        if n['use'] == 'usual':
-                            last_name = n['family']
-                            first_name = n['given'][0]
-                        if last_name or first_name:
-                            user.first_name = first_name
-                            user.last_name = last_name
-                            user.save()
-    except (UpstreamServerException, exceptions.NotFound):
-        logger.error("Failed to connect Beneficiary "
-                     "to FHIR")
+        if first_name == "" or last_name == "":
+            first_name, last_name = extract_beneficiary_names(backend_data)
+    except exceptions.NotFound:
+        fhir_id = None
 
+    fhir_source = get_resourcerouter()
+
+    user = create_beneficiary_record(username=subject,
+                                     user_id_hash=hicn_hash,
+                                     fhir_id=fhir_id,
+                                     fhir_source=fhir_source,
+                                     first_name=first_name,
+                                     last_name=last_name,
+                                     email=email)
     return user
 
 
+def create_beneficiary_record(username=None,
+                              user_id_hash=None,
+                              fhir_id=None,
+                              fhir_source=None,
+                              first_name=None,
+                              last_name=None,
+                              email=None):
+    with transaction.atomic():
+        user = User(username=username,
+                    first_name=first_name,
+                    last_name=last_name,
+                    email=email)
+        user.set_unusable_password()
+        user.save()
+        Crosswalk.objects.create(user=user,
+                                 fhir_source=fhir_source,
+                                 user_id_hash=user_id_hash,
+                                 fhir_id=fhir_id)
```

Comment:
Can fhir_id be None? If so, I question that we should allow this. In other words, I don't see the need for authenticate_crosswalk to set it.

Reply:
In the current requirements it can be None, there are plans to investigate if there's a path to changing this.

Can you give an example of the change you're suggesting here?

Reply:
Maybe asserts that hicn_hash and fhir_id are not None. Change the DB to not allow nulls in these fields.

Reply:
:100: that work's in progress here https://github.com/CMSgov/bluebutton-web-server/pull/772

At the moment `fhir_id` has to remain nullable.

Reply:
I've added some assertions to this function as well as what's coming in #772